flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/resources/KubernetesSharedInformer.java

```diff
@@ -0,0 +1,285 @@
+
+    private final ExecutorService watchingExecutor;
+
+    private final Object processorsLock = new Object();
+
+    @GuardedBy("processorsLock")
+    private final Map<String, Processor> watchingProcessors = new HashMap<>();
+
+    public KubernetesSharedInformer(
+            NamespacedKubernetesClient client,
+            Class<T> apiTypeClass,
+            Class<TList> apiListTypeClass,
+            Map<String, String> labels,
+            Function<T, R> eventWrapper) {
+        Preconditions.checkArgument(
+                !CollectionUtil.isNullOrEmpty(labels), "Labels must not be null or empty");
+        this.client = client;
+        SharedInformerFactory informerFactory =
+                client.informers(
+                        Executors.newSingleThreadExecutor(
+                                new ExecutorThreadFactory("KubernetesClient-Informer")));
+        informerFactory.withLabels(labels);
+        this.sharedIndexInformer =
+                informerFactory.sharedIndexInformerFor(apiTypeClass, apiListTypeClass, 0);
+        this.sharedIndexInformer.addEventHandler(
+                new ResourceEventHandler<T>() {
+                    @Override
+                    public void onAdd(T obj) {
+                        notify(obj.getMetadata().getName());
+                    }
+
+                    @Override
+                    public void onUpdate(T oldObj, T newObj) {
+                        notify(newObj.getMetadata().getName());
+                    }
+
+                    @Override
+                    public void onDelete(T obj, boolean deletedFinalStateUnknown) {
+                        notify(obj.getMetadata().getName());
+                    }
+
+                    private void notify(String name) {
+                        final String resourceKey = getResourceKey(name);
+                        synchronized (processorsLock) {
+                            Optional.ofNullable(watchingProcessors.get(resourceKey))
+                                    .ifPresent(processor -> processor.triggerEvent(resourceKey));
+                        }
+                    }
+                });
+
+        this.watchingExecutor =
+                Executors.newCachedThreadPool(
+                        new ExecutorThreadFactory("KubernetesSharedInformer-Watching-Processor"));
+        this.eventWrapper = eventWrapper;
+    }
+
+    @Override
+    public Watch watch(String name, WatchCallbackHandler<R> handler) {
+        final String resourceKey = getResourceKey(name);
+        synchronized (processorsLock) {
+            Processor existed = watchingProcessors.get(resourceKey);
+            if (existed != null) {
+                final Optional<UUID> uuid = existed.registerListener(handler);
+                if (uuid.isPresent()) {
+                    return () -> existed.unregisterListener(uuid.get());
+                }
+            }
+            Processor processor = new Processor(resourceKey);
```

Comment:
nit: Use `final` as much as possible.

Maybe we could also have a comment here. Something like "Create a new one if the processor is stopped or does not existed".